rust/kernel/error.rs

```diff
@@ -263,3 +244,60 @@ pub(crate) fn from_kernel_err_ptr<T>(ptr: *mut T) -> Result<*mut T> {
     }
     Ok(ptr)
 }
+
+/// Transform a kernel integer result to a [`Result`].
+///
+/// Some kernel C API functions return a result in the form of an integer:
+/// zero if ok, a negative errno on error. This function converts such a
+/// return value into an idiomatic [`Result`].
+///
+/// # Safety
+///
+/// `retval` must be non-negative or a valid negative errno (i.e. `retval` must
+/// be in `[-MAX_ERRNO..]`).
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// let ret = unsafe { bindings::misc_register(&mut this.mdev) };
+/// // SAFETY: `misc_register` returns zero on success, or a valid
+/// // negative errno on failure.
+/// unsafe { from_kernel_int_result(ret)?; }
+/// ```
+pub(crate) unsafe fn from_kernel_int_result(retval: c_types::c_int) -> Result {
```

Comment:
So this one loses the error code, but the unsigned one does not -- is it on purpose?

Reply:
Yes it is. This one deals with kernel C functions that return zero on success, or negative errno. They're not supposed to return anything except "success!". But, if something went wrong and the function did return something `>0`, I would not want that value to end up inside an `Error` ! So I treat anything non-negative as success.

Reply:
> They're not supposed to return anything except "success!".

Ah, yes, I misread, my bad!

> But, if something went wrong and the function did return something `>0`

Hmm... that would be a bug in the function, no? i.e. if we are using `from_kernel_int_result` it is because the function we call should never return something `>0`, i.e. there is no "return value" when they succeed.

Perhaps we can clarify the docs a bit:

"Use this function when the C function only returns `0` on success or negative on error. If the C function returns a useful value on the happy path, use `from_kernel_int_result_uint` instead."

And for the other one:

"Use this function when the C function returns a useful, positive value on success, or negative on error. If the C function only returns `0` on success, use `from_kernel_int_result` instead."